src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,609 @@
+
+    double dp;
+    double dr;
+
+    if (last_pitch > CV_PI/2 && pitch < -CV_PI/2) {
+        dp = (pitch + 2*CV_PI - last_pitch);
+    } else if (last_pitch < -CV_PI/2 && pitch > CV_PI/2) {
+        dp = (pitch - last_pitch - 2*CV_PI);
+    } else {
+        dp = (pitch - last_pitch);
+    }
+
+    if (last_roll > CV_PI/2 && roll < -CV_PI/2) {
+        dr = (roll + 2*CV_PI - last_roll);
+    } else if (last_roll < -CV_PI/2 && roll > CV_PI/2) {
+        dr = (roll - last_roll - 2*CV_PI);
+    } else {
+        dr = (roll - last_roll);
+    }
+
+    double dpitch_dt = dp / dt;
+    double droll_dt = dr / dt;
+
+    // Observed velocity in camera frame due to rotation
+    //
+    // Positive dpdt means that the x vector rotates down, which means we think
+    // we're moving in the -x direction
+    //
+    // Positive drdt means that the y vector rotates down, which means we think
+    // we're moving in the -y direction
+    //
+    // Note again that cos(roll) is negative, because the z axis is down
+    cv::Point2f correction_vel;
+    correction_vel.x = distance_to_plane
+                     * -dpitch_dt
+                     / std::cos(pitch);
+    correction_vel.y = distance_to_plane
+                     * -droll_dt
+                     / -std::cos(roll);
+
+    // Actual velocity in level camera frame (i.e. camera frame if our pitch
+    // and roll were zero)
+    Eigen::Vector3d corrected_vel(
+        velocity_uncorrected.x - correction_vel.x,
+        velocity_uncorrected.y - correction_vel.y,
+        0.0);
+
+    last_pitch = pitch;
+    last_roll = roll;
+
+    // Calculate covariance
+    Eigen::Matrix3d covariance = Eigen::Matrix3d::Zero();
+    covariance(0, 0) = std::pow(flow_estimator_settings_.variance_scale
+                              * dpitch_dt, 2.0)
+                      + flow_estimator_settings_.variance;
+    covariance(1, 1) = std::pow(flow_estimator_settings_.variance_scale
+                              * droll_dt, 2.0)
+                      + flow_estimator_settings_.variance;
+
+    // Rotation matrix from level camera frame to level_quad
+    Eigen::Matrix3d rotation_matrix;
+    rotation_matrix = Eigen::AngleAxisd(M_PI, Eigen::Vector3d::UnitX())
+                    * Eigen::AngleAxisd(-yaw, Eigen::Vector3d::UnitZ());
+
+    // Get velocity and covariance in level_quad frame
+    Eigen::Vector3d level_quad_vel = rotation_matrix * corrected_vel;
+    Eigen::Matrix3d level_quad_covariance = rotation_matrix
+                                          * covariance
+                                          * rotation_matrix.inverse();
+
+    // Correct for movement of camera relative to level_quad
+    geometry_msgs::PointStamped curr_pos, last_pos;
+    tf2::doTransform(curr_pos, curr_pos, current_camera_to_level_quad_tf_);
+    tf2::doTransform(last_pos, last_pos, last_camera_to_level_quad_tf_);
+    double camera_relative_vel_x = (curr_pos.point.x - last_pos.point.x) / dt;
```

Comment:
Same as comment above about using the kalmann filter to get this information. We could get the angular velocity and then use it to calculate the offset vector.

Reply:
Again, we care about the average velocity over the interval between frames, not the instantaneous velocity